src/main/java/org/spongepowered/common/world/FakePlayer.java

```diff
@@ -0,0 +1,195 @@
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.event.tracking.PhaseTracker;
+import org.spongepowered.common.event.tracking.phase.plugin.PluginPhase;
+import org.spongepowered.common.registry.provider.DirectionFacingProvider;
+import org.spongepowered.common.util.VecHelper;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A marker to state that this is a fake player.
+ */
+public interface FakePlayer {
+
+    Controller controller = new Controller(SpongeImpl.getGame().createFakePlayerFactory());
+
+    public class Factory {
+
+        private final Map<UUID, EntityPlayerMP> players = new MapMaker().weakValues().makeMap();
+
+        public EntityPlayerMP getPlayer(WorldServer world, com.mojang.authlib.GameProfile profile) {
+            EntityPlayerMP player = players.get(profile.getId());
+            if (player == null) {
+                players.put(profile.getId(), player = new SimulatedPlayer(world, profile));
+            }
+            return player;
+        }
+
+    }
+
+    public static class Controller {
+
+        private final Factory factory;
+
+        Controller(Factory factory) {
+            this.factory = factory;
+        }
+
+        private EntityPlayerMP setUpPlayer(GameProfile profile, World world, int x, int y, int z, ItemStack itemInHand) {
+            EntityPlayerMP player = factory.getPlayer((WorldServer) world, (com.mojang.authlib.GameProfile) profile);
+            player.world = (WorldServer) world;
+            player.interactionManager.setWorld((WorldServer) world);
+            player.posX = x;
+            player.posY = y;
+            player.posZ = z;
+            player.onGround = true;
+            net.minecraft.item.ItemStack stack = (net.minecraft.item.ItemStack) itemInHand;
+            if (stack == null) {
+                stack = net.minecraft.item.ItemStack.EMPTY;
+            }
+            player.inventory.setItemStack(stack);
+            player.inventory.mainInventory.set(player.inventory.currentItem, stack);
+
+            // Set up cause capture
+            PluginPhase.State.FAKE_PLAYER.createPhaseContext()
+                    .addCaptures()
+                    .buildAndSwitch();
+            Sponge.getCauseStackManager().addContext(EventContextKeys.PLAYER_SIMULATED, profile);
+            return player;
+        }
+
+        private static void tearDownPlayer(EntityPlayerMP player) {
+            // Tear down cause capture
+            PhaseTracker.getInstance().completePhase(PhaseTracker.getInstance().getCurrentState());
+
+            player.inventory.clear();
+            player.world = null;
+            player.interactionManager.setWorld(null);
+        }
+
+        public boolean hit(World world, int x, int y, int z, Direction side, GameProfile profile) {
+            EntityPlayerMP player = setUpPlayer(profile, world, x, y, z, null);
```

Comment:
One thing I can imagine looking at this PR is that we can make this entirely contained within good `try catch` blocks like so:
```java
try (SpongeFakePlayer player = setUpPlayer(profile, world, x, y, z null)) {
  EnumFacing.....
  // yadda yadda yadda
}

```

In the sense that the SpongeFakePlayer would be `AutoCloseable` to call `tearDownPlayer` properly. This can also play into pushing onto the `CauseStackManager` for the game profile object etc.